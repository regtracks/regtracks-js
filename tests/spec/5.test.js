const Track = require('../../index');

describe('tests section 5 of the specification', () => {
  test('section 5.1: symbols', () => {
    let schema = String.raw`
      @valid_23b2
      ||
      any a to z
      ||

      @againValid
      a, b, c, digit, space
    `;

    const track = new Track(schema);
    expect(track.parser.tree.tree).toMatchSnapshot();

    let illFormedSchema = String.raw`
      @name_not_before
      @symbol2
      ||
      any a to z
      ||
    `;

    expect(() => new Track(illFormedSchema)).toThrow();
  });

  test('sections 5.2, 5.4-5: bad symbol names', () => {
    let illFormedSchema = String.raw`
      @pattern
      ||
      @not_here
      any a to z
      ||
    `;

    let illFormedSchema2 = String.raw`
      @a
      ||
      any a to z
      ||
    `;

    let illFormedSchema3 = String.raw`
      @5nonum
      ||
      any a to z
      ||
    `;

    expect(() => new Track(illFormedSchema)).toThrow();
    expect(() => new Track(illFormedSchema2)).toThrow();
    expect(() => new Track(illFormedSchema3)).toThrow();
  });

  test('section 5.6: same symbol name more than once', () => {
    let illFormedSchema = String.raw`
      @pattern
      ||
      any a to z
      ||

      @pattern
      ||
      (:)
      ||
    `;

    expect(() => new Track(illFormedSchema)).toThrow();
  });

  test('section 5.3: reserved symbol names', () => {
    const reserved = [
      'after',
      'as',
      'any',
      'anything',
      'digit',
      'end',
      'forever',
      'none',
      'optionally',
      'or',
      'space',
      'start',
      'times',
      'to',
      'whitespace',
    ];

    for (let name of reserved) {
      let illFormedSchema = String.raw`
        @${ name }
        ||
        any a to z
        ||
      `;
      expect(() => new Track(illFormedSchema)).toThrow();
    }
  });
});
