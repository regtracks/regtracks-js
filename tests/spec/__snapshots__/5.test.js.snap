// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests section 5 of the specification section 5.1: symbols 1`] = `
Object {
  "againValid": Object {
    "matches": Array [
      Object {
        "matches": "a",
        "type": "literal_criterion",
      },
      Object {
        "matches": "b",
        "type": "literal_criterion",
      },
      Object {
        "matches": "c",
        "type": "literal_criterion",
      },
      Object {
        "matches": "digit",
        "type": "symbol_criterion",
      },
      Object {
        "matches": "space",
        "type": "symbol_criterion",
      },
    ],
    "name": "againValid",
    "type": "charset",
  },
  "valid_23b2": Object {
    "name": "valid_23b2",
    "root": Object {
      "children": Array [
        Object {
          "collectionIdent": undefined,
          "forks": false,
          "inclusive": true,
          "matches": Array [
            Object {
              "matches": /\\[a-z\\]/,
              "type": "character_range_criterion",
            },
          ],
          "notMatch": false,
          "optional": false,
          "repeat": undefined,
          "type": "matching_rule",
        },
      ],
      "collectionIdent": undefined,
      "forks": false,
      "notMatch": false,
      "optional": false,
      "parent": undefined,
      "repeat": undefined,
      "type": "block",
    },
    "type": "pattern",
  },
}
`;
